core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageJNIConfigStep.java

```diff
@@ -47,6 +49,13 @@ void generateJniConfig(BuildProducer<GeneratedResourceBuildItem> jniConfig,
             root.add(json);
         }
 
+        // Add raw, unescaped, unprocessed nodes for user's fine-grained control.
+        for (JniRuntimeAccessJSONBuildItem jniRuntimeAccessJSONBuildItem : jniRuntimeJSON) {
+            for (String j : jniRuntimeAccessJSONBuildItem.jsonNodes) {
+                root.addRaw(j);
+            }
+        }
+
```

Comment:
@Karm why did you choose to go with raw JSON instead of doing something similar to what we do for reflection?
It feels fragile to me.

https://github.com/quarkusio/quarkus/blob/da625e30ffe4a4875c023b69ca41c133bfacf50a/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageReflectConfigStep.java#L92-L106

I can prepare a commit on top of this PR to implement the latter if you agree.

Reply:
> How exactly is it more or less fragile than calling method by names one by one?

I find it more fragile in the sense that we rely on hand-crafted raw json. It's not really a blocker for me but I would like to see what others thing as well.

> Please, postpone modifications until I push the JDK 20 port as thre might be more changes....

:+1:

Reply:
cc usual suspects @gsmet @geoand ^^

Reply:
I'd really like to avoid the raw json please

Reply:
@geoand @zakkak @gsmet  I am not sure I am making a good job at painting the full picture. Let me make an attempt to defend this approach :-D

* You cannot use the current level of granularity, i.e. you have to have a way to mark single methods or even single fields.
* We are using the JSON already, exactly the same format. We are generating it from what extension developers mark with respective builder items.

Now, you spent considerable time with native agent and doing manual work in gdb to arrive at a JSON *x*.

Why would you re-type that JSON snippets native agent got you into some new Quarkus specific API....only to generate JSON *y* (of the exactly the same format you already had in the first place)?

It feels like you have one JSON in your left hand, the same type of JSON in your right hand and somehow you are re-processing it in the middle uselessly.

I propose that we leave this "raw" or "advance" option open so as one can directly inject JSON without having Quarkus re-process it again.

The alternative is that I create an API that allows you to specify single methods and single fields, some arrays of methods and arrays of fields and re-type everything native-agent generates for me...all of it again serialized into the same JSON.

## JSON

You already have some JSON snippets from your native agent, so you make it correct (complete) for Quarkus and copy paste into a file:

```java
@BuildStep
JniRuntimeAccessJSONBuildItem setupAWTInit() throws URISyntaxException, IOException {
    final String jsonArray = Files.readString(Paths.get(getClass().getResource("/jni-json-snippet.json").toURI()));
    // Strip leading [ and trailing ] as we are injecting these elements into an already existing JSON array.
    return new JniRuntimeAccessJSONBuildItem(jsonArray.substring(1, jsonArray.length() - 1));
}
```

The file might look like this:

```json
[
  {
    "name": "java.lang.System",
    "methods": [
      {
        "name": "load",
        "parameterTypes": [
          "java.lang.String"
        ]
      },
      {
        "name": "setProperty",
        "parameterTypes": [
          "java.lang.String",
          "java.lang.String"
        ]
      }
    ]
  },
  {
    "name": "sun.awt.SunToolkit",
    "methods": [
      {
        "name": "awtLock",
        "parameterTypes": []
      },
      {
        "name": "awtLockNotify",
        "parameterTypes": []
      },
      {
        "name": "awtLockNotifyAll",
        "parameterTypes": []
      },
      {
        "name": "awtLockWait",
        "parameterTypes": [
          "long"
        ]
      },
      {
        "name": "awtUnlock",
        "parameterTypes": []
      }
    ],
    "fields": [
      {
        "name": "AWT_LOCK"
      },
      {
        "name": "AWT_LOCK_COND"
      }
    ]
  },
  {
    "name": "sun.awt.X11.XErrorHandlerUtil",
    "methods": [
      {
        "name": "init",
        "parameterTypes": [
          "long"
        ]
      }
    ]
  },
  {
    "name": "sun.awt.X11.XToolkit",
    "allDeclaredFields": true
  },
  {
    "name": "java.lang.Thread",
    "methods": [
      {
        "name": "yield",
        "parameterTypes": []
      }
    ]
  }
]
```
Those nodes are then injected into the bigger jni conf JSON generated by Quarkus.
While doing this work, you are inspecting the generated JSON files anyway, to make sure you are not chasing ghosts.

## API

You have to either manually type stuff from JSON file native agent gave you or have s script that transforms that into this Java code.
An example of an API I came up with minutes ago while eating lunch. Perhaps it could be made more user friendly:

```java
List<Klass> c...
c.add(new KlassBuilder("java.lang.System")
        .addMethods(Map.of("load", List.of("java.lang.String"),
                "setProperty", List.of("java.lang.String", "java.lang.String")))
        .build()
);
c.add(new KlassBuilder("sun.awt.SunToolkit")
        .addMethods(Map.of("awtLock", Collections.emptyList(),
                 "awtLockNotify", Collections.emptyList(),
                "awtLockNotifyAll", Collections.emptyList(),
                "awtLockWait", List.of("long"),
                "awtUnlock", Collections.emptyList()))
        .addFields(List.of("AWT_LOCK", "AWT_LOCK_COND"))
        .build()
);
c.add(new KlassBuilder("sun.awt.X11.XErrorHandlerUtil")
        .addMethod(Map.of("init", List.of("long")))
        .build()
);
c.add(new KlassBuilder("sun.awt.X11.XToolkit")
        .addAllFields()
        .build()
);
c.add(new KlassBuilder("java.lang.Thread")
        .addMethod(Map.of("yield", List.of()))
        .build()
);
```

The thing is that Quarkus then takes this, and generates *the same JSON you already had*.


## All in all

I would very well understand any opposition to "raw JSON" snippets prior to https://github.com/quarkusio/quarkus/pull/29886,
but we are already using JSON config files now.

I hope to be able to contribute to native agent at some point in not so distant future so as it could take over this for Quarkus, at least for JNI access.

If you guys feel like it's a better option, I'll provide the aforementioned API and a jbang script that generates that Java code.

Reply:
@Karm please see https://github.com/Karm/quarkus/pull/106

Reply:
Adopted.